CUDADataFormats/HGCal/interface/HGCRecHitSoA.h

```diff
@@ -0,0 +1,27 @@
+#ifndef CUDADataFormats_HGCal_HGCRecHitSoA_h
+#define CUDADataFormats_HGCal_HGCRecHitSoA_h
+
+class HGCRecHitSoA {
+public:
+  float *energy_;       //calibrated energy of the rechit
+  float *time_;         //time jitter of the UncalibRecHit
+  float *timeError_;    //time resolution
+  uint32_t *id_;        //rechit detId
+  uint32_t *flagBits_;  //rechit flags describing its status (DataFormats/HGCRecHit/interface/HGCRecHit.h)
+  uint8_t *son_;        //signal over noise
```

Comment:
@makortel What do you then suggest? From the code's point of view, the data is filled in the GPU (so cannot be declared as const), and the pointers themselves are accessed in ```RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalProducerMemoryWrapper.cc```, and cannot be const as well.
Should I change the data structure, or just the way that data is being accessed? Is something like [this](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitProducer.cc#L70-L88) better?

Reply:
> Is something like [this](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitProducer.cc#L70-L88) better?

Taking Event products by const reference or const pointer would be a step forward, but unfortunately would not be enough. The difference to the code in the link is that when copied those objects are do deep copies, whereas this struct does a shallow copy.

In a bit similar situation https://github.com/cms-sw/cmssw/pull/32039 used `edm::propagate_const`. But the same technique would not work here because raw pointers are copyable and so it would be easy to get a non-const copy of a const object that would allow circumventing `edm::propagate_const`  (in #32039 the contained pointers were always `unique_ptr` that can not be copied; the structs of `unique_ptr`s were also kept in host memory).

One option would be to use `edm::propagate_const` and disallow copying of `HGCRecHitSoA`. But then you would not be able to pass `HGCRecHitSoA` to kernels by value (which I see you're doing).

Alternative would be to define a separate "const view" struct along
```cpp
struct ConstHGCRecHitSoA {
  float const *energy_;
  ...
  uint32_t nbytes_;
  ...
};
```
and use that such that the object that is put in the Event (and owns the memory and is non-copyable) returns `HGCRecHitSoA` when it is non-const, and `ConstHGCRecHitSoA` when it is const.

(what we'd really need is a nice abstraction that would do all these for you)

Reply:
From what I understand, I can either stop passing the products by value or else define the abstraction you mentioned. 
As for the former, I was "forced" to do it since, in order to preserve the (in my opinion nice) SoA structure, the addresses stored in the SoA's pointers must be known within the GPU (otherwise I lose access to their allocated memory). 
I could always allocate a different block of memory for each SoA pointer, and pass those by value, but I understood one always wants to reduce the number of GPU memory allocations. I seem to recall this was the approach followed for the PixelTracks, which was considered not to be the best option.
A viable alternative would be to use CUDA Unified Memory. After obtaining my *constant* SoA product, I would only need to pass it by reference, and the addresses of its pointers would be known also on the GPU side. However, I do not how I should follow this approach in the context of the ```cms::cuda::make_*``` utilities. In addition, I am wondering if you happen to be aware of any third option to avoid passing the product by value.
Instead, regarding the approach you proposed, does it mean a new 'const view' would have to be created every time a const SoA is requested? How would the object know which version to return without some user input?
Finally, when considering the two major possibilities discussed above, which one do you prefer?

Reply:
> As for the former, I was "forced" to do it since, in order to preserve the (in my opinion nice) SoA structure, the addresses stored in the SoA's pointers must be known within the GPU (otherwise I lose access to their allocated memory).

I agree this approach of passing a struct of non-owning pointers has nice elements, and IIRC it has not been tested in practice yet (even if we've talked about it).

> I could always allocate a different block of memory for each SoA pointer, and pass those by value, but I understood one always wants to reduce the number of GPU memory allocations. I seem to recall this was the approach followed for the PixelTracks, which was considered not to be the best option.

To leading order the "View" interface and whether allocate each column separately or not are orthogonal. E.g. the [`TrackingRecHit2DHeterogeneous`](https://github.com/cms-sw/cmssw/blob/master/CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h) allocates multiple columns at a time (32-bit and 16-bit columns separately) and uses a [View](https://github.com/cms-sw/cmssw/blob/master/CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DSOAView.h). In principle the `TrackingRecHit2DSOAView` should be declared non-copyable, but in practice the risks with the copy ability have been lower than here because that View is allocated in device memory (so trying to copy in host code would lead to runtime error), and the View is passed around by pointer.

> A viable alternative would be to use CUDA Unified Memory.

I would avoid unified memory, because it will have performance implications (evaluation is still ongoing, but work so far suggests that with pixel reconstruction we're talking about a difference of factor(s) rather than percentages, but YMMV), and the data structure is not that complicated.

I also didn't understand what you would be able save compared to explicit memory than one allocation and the explicit copy, which, to me, are not that big deal compared to assignment of column pointers to the memory block. 

> Instead, regarding the approach you proposed, does it mean a new 'const view' would have to be created every time a const SoA is requested?

Yes. But setting a handful of pointers is relatively simple operation. You could also hold one copy of the View internally and just copy it when requested.

> How would the object know which version to return without some user input?

When the object is const, return "const view", when it is non-const, return "non-const view", along
```cpp
class HGCRecHitGPUProduct {
  ...
  HGCRecHitSoA get() { return ...; }
  ConstHGCRecHitSoA get() const { return ...; }
}
```

> Finally, when considering the two major possibilities discussed above, which one do you prefer?

I'd probably try the "const view" approach, and passing it to the kernel by value, or resort to the "allocate view on the device memory" approach that's used in the pixel code.